File: Map.py
import folium
from folium.plugins import LocateControl, MousePosition, Geocoder
import json

# Coordinates that create a square around Abuja
min_lon, max_lon = 7.15, 7.8
min_lat, max_lat = 8.8, 9.2

# Base map, location of the mid-point & zoom
map = folium.Map(
    max_bounds=True,
    location=[9.072264, 7.491302],
    zoom_start=12,
    min_zoom=12,
    max_zoom=18,
    min_lat=min_lat,
    max_lat=max_lat,
    min_lon=min_lon,
    max_lon=max_lon,
    # Delete comment from the 2 rows under to use different map layout
    # tiles="https://{s}.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
    # attr='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors, Tiles style by <a href="https://www.hotosm.org/" target="_blank">Humanitarian OpenStreetMap Team</a> hosted by <a href="https://openstreetmap.fr/" target="_blank">OpenStreetMap France</a>',
)


folium.CircleMarker([max_lat, min_lon], tooltip="Upper Left Corner").add_to(map)
folium.CircleMarker([min_lat, min_lon], tooltip="Lower Left Corner").add_to(map)
folium.CircleMarker([min_lat, max_lon], tooltip="Lower Right Corner").add_to(map)
folium.CircleMarker([max_lat, max_lon], tooltip="Upper Right Corner").add_to(map)

# Accessing the json where busstops are located
with open("POI.json") as file:
    data = json.load(file)

# Feature group
bus_stops = folium.FeatureGroup(name="Bus stops")

# Go through all the stops in the json and add all of them to the map as markers
for i in range(len(data["bus_stops"])):

    bus_stop_name = data["bus_stops"][i]["name"]
    html_with_name = f'<h1 id="bus_stop_name">{bus_stop_name}</h1><br>Timetable for the bus stop:<p><code>Insert timetable here...</code></p>'

    # Create a Popup object with the HTML content
    popup = folium.Popup(html_with_name, max_width=400)

    # Color and icon style of the marker
    color = data["bus_stops"][i]["color"]
    icon = data["bus_stops"][i]["icon"]

    # Add marker to the feature group
    bus_stops.add_child(
        folium.Marker(
            location=[
                data["bus_stops"][i]["latitude"],
                data["bus_stops"][i]["longitude"],
            ],
            popup=popup,
            icon=folium.Icon(color=color, icon=icon, prefix="fa"),
        )
    )
    # Add feature group to the map
    map.add_child(bus_stops)

# Gps tracking request through browser
location = LocateControl().add_to(map)

# Add current mouse coordinates
mouse = MousePosition().add_to(map)

# Add layercontrol UI
layer_control = folium.LayerControl().add_to(map)

# Add search plugin
search = Geocoder().add_to(map)

# Renders the map without creating html file
# map.show_in_browser()

# Rendering + alternative methods commmented down below
map.save("map.html")

File: public/js/map.js
// Create base map and set the location to Abuja, Nigeria
var map = L.map("map", {
    center: [9.05, 7.491302],
    crs: L.CRS.EPSG3857,
    maxBounds: [
        [8.8, 7.15],
        [9.2, 7.8],
    ],
    zoom: 14,
    zoomControl: true,
    preferCanvas: false,
});

// Choose tiles that the map uses and set min and max zoom
var tileLayer = L.tileLayer("https://tile.openstreetmap.org/{z}/{x}/{y}.png", {
    maxZoom: 19,
    minZoom: 11,
    attribution:
        '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>',
}).addTo(map);

// Create layers for the different busstops routes
var airport_to_city = L.markerClusterGroup({
    disableClusteringAtZoom: 13,
}).addTo(map);
var route_2 = L.markerClusterGroup({ disableClusteringAtZoom: 13 }).addTo(map);
var route_3 = L.markerClusterGroup({ disableClusteringAtZoom: 13 }).addTo(map);

// Create busIcon
var busIconRoute1 = L.icon({
    iconUrl: "../img/b.png",
    iconSize: [80, 80], // Set the size of the icon
    iconAnchor: [15, 30], // Set the anchor point of the icon
    popupAnchor: [0, -30], // Set the popup anchor point
});

var busIconRoute2 = L.icon({
    iconUrl: "../img/a.png",
    iconSize: [80, 80], // Set the size of the icon
    iconAnchor: [15, 30], // Set the anchor point of the icon
    popupAnchor: [0, -30], // Set the popup anchor point
});

var busIconRoute3 = L.icon({
    iconUrl: "../img/e.png",
    iconSize: [80, 80], // Set the size of the icon
    iconAnchor: [15, 30], // Set the anchor point of the icon
    popupAnchor: [0, -30], // Set the popup anchor point
});

// Fetch bus stop data from JSON file
fetch("../databases/POI.json")
    .then((response) => response.json())
    .then((data) => {
        // Process the JSON data and create markers
        // For example:
        data.bus_stops.forEach((busStop) => {
            var popup = L.popup({ maxWidth: 400, maxHeight: 300 });
            popup.setContent(
                `<div id="popup" style="width: 100.0%; height: 100.0%;"><h1>${busStop.id}. ${busStop.name}</h1><br>Timetable for the bus stop:<p>${busStop.timetable}</p></div>`
            );

            var busIcon;
            if (busStop.route === 1) {
                busIcon = busIconRoute1;
            } else if (busStop.route === 2) {
                busIcon = busIconRoute2;
            } else if (busStop.route === 3) {
                busIcon = busIconRoute3;
            }

            var bus_marker = L.marker([busStop.latitude, busStop.longitude], {
                icon: busIcon,
            });

            if (busStop.route === 1) {
                bus_marker.addTo(airport_to_city);
            } else if (busStop.route === 2) {
                bus_marker.addTo(route_2);
            } else if (busStop.route === 3) {
                bus_marker.addTo(route_3);
            }

            bus_marker.bindPopup(popup);
        });
    })
    .catch((error) => console.error("Error loading JSON file:", error));

// Add length ruler to bottom left corner
var control_scale = L.control.scale().addTo(map);

// Create baselayer for bus stops
var baseLayers = {
    "Airport to City": airport_to_city,
    "Route 2": route_2,
    "Route 3": route_3,
};

// Add layer control so that markers can be toggled on & off
var layer_control = L.control.layers(null, baseLayers).addTo(map);

// Add geocoder control (Search bar)
var geocoder = L.Control.geocoder({
    position: "topright",
    expand: "hover",
}).addTo(map);

// Add current mouse coordinates to the map
var mousePosition = L.control
    .mousePosition({ position: "bottomright" })
    .addTo(map);
